jxl/src/entropy_coding/ans.rs

```diff
@@ -2,6 +2,410 @@
+                log_bucket_size,
+                bucket_mask: (1 << log_bucket_size) - 1,
+                single_symbol: Some(single_sym_idx as u32),
+            });
+        }
+
+        let mut buckets: Vec<_> = dist
+            .into_iter()
+            .enumerate()
+            .map(|(i, dist)| WorkingBucket {
+                dist,
+                alias_symbol: if i < alphabet_size { i as u16 } else { 0 },
+                alias_offset: 0,
+                alias_cutoff: dist,
+            })
+            .collect();
+
+        let mut underfull = Vec::new();
+        let mut overfull = Vec::new();
+        for (idx, &WorkingBucket { dist, .. }) in buckets.iter().enumerate() {
+            match dist.cmp(&bucket_size) {
+                std::cmp::Ordering::Less => underfull.push(idx),
+                std::cmp::Ordering::Equal => {}
+                std::cmp::Ordering::Greater => overfull.push(idx),
+            }
+        }
+        while let (Some(o), Some(u)) = (overfull.pop(), underfull.pop()) {
+            let by = bucket_size - buckets[u].alias_cutoff;
+            buckets[o].alias_cutoff -= by;
+            buckets[u].alias_symbol = o as u16;
+            buckets[u].alias_offset = buckets[o].alias_cutoff;
+            match buckets[o].alias_cutoff.cmp(&bucket_size) {
+                std::cmp::Ordering::Less => underfull.push(o),
+                std::cmp::Ordering::Equal => {}
+                std::cmp::Ordering::Greater => overfull.push(o),
+            }
+        }
+
+        let buckets = buckets
+            .iter()
+            .enumerate()
+            .map(|(idx, bucket)| {
+                if bucket.alias_cutoff == bucket_size {
+                    Bucket {
+                        dist: bucket.dist,
+                        alias_symbol: idx as u8,
+                        alias_offset: 0,
+                        alias_cutoff: 0,
+                        alias_dist_xor: 0,
+                    }
+                } else {
+                    Bucket {
+                        dist: bucket.dist,
+                        alias_symbol: bucket.alias_symbol as u8,
+                        alias_offset: bucket.alias_offset - bucket.alias_cutoff,
+                        alias_cutoff: bucket.alias_cutoff as u8,
+                        alias_dist_xor: bucket.dist ^ buckets[bucket.alias_symbol as usize].dist,
+                    }
+                }
+            })
+            .collect();
+
+        Ok(Self {
+            buckets,
+            log_bucket_size,
+            bucket_mask: (1 << log_bucket_size) - 1,
+            single_symbol: None,
+        })
+    }
+
+    fn read_u8(bitstream: &mut BitReader) -> Result<u8> {
+        Ok(if bitstream.read(1)? != 0 {
+            let n = bitstream.read(3)?;
+            ((1 << n) + bitstream.read(n as usize)?) as u8
+        } else {
+            0
+        })
+    }
+
+    fn read_prefix(br: &mut BitReader) -> Result<u16> {
+        Ok(match br.read(3)? {
```

Comment:
I assume this is equivalent to the lookup table in libjxl and the spec - any reason not to do just that?

Reply:
This is one of the oldest code in jxl-oxide, and I think I wasn't aware of doing prefix code with lookup tables at that time.